k8s/node-init: add node-init script to automatically restart pods managed by kubenet on GKE

aanm opened the pull request:
Please read per commit basis, I couldn't find a better solution than having 2 node-init scripts.

```release-note
k8s/node-init: add node-init script to automatically restart pods managed by kubenet on GKE
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7486)
<!-- Reviewable:end -->

tgraf commented on examples/kubernetes/node-init/node-init.yaml:
```diff
@@ -45,41 +52,53 @@ spec:
+                echo "Restarting kubenet managed pods"
+                if grep -q 'docker' /etc/crictl.yaml; then
+                  # Works for COS, ubuntu
+                  for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do docker rm -f $(cat $f) || true; done
```
This seems very aggressive. Any reason not to do this via the apiserver? At the very least I would hide this behind an option. This could have a lot of consequences that we don't want.

aanm replied:
>At the very least I would hide this behind an option.

Yeah I was thinking the same.

>Any reason not to do this via the apiserver?

The user will have to remember to do that step in the kube-apiesrver every time a new node is provisioned.

aanm committed:
kubernetes/node-init: run cilium-node-init on any tainted node

the node-init script is essential for the cluster to run with cilium
therefore it should run on any tainted node.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/node-init: run cilium-node-init in hostNetwork

cilium-node-init does not depend on any CNI plugin to be setup so we can
run it on hostNetwork.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/node-init: do not run script on an already setup node

If the DaemonSet is accidentally removed it can be re-added without
fearing the script will be executed again.

However, if the node is restarted, or upgraded, the node-init script
might need to be ran again which node-init will detected based on the
existence of a file stored in the `/tmp` directory.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/node-init: Install cilium cni config before restart kubelet

If the CNI configuration is installed afterwards there might be a race
where scheduled pods might be managed by kubenet instead of Cilium.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/node-init: add more aggressive node-init script

As node-init script is set with an env variable we can not use another
environment variable to enable or disable the automatic deletion of pods
managed by kubenet.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/node-init: delete cilium running before kubelet restart

We still need to delete Cilium in case it started before we have changed
kubelet configuration. This prevents Cilium from pre-allocating all IP
addresses managed by cbr0.

Signed-off-by: André Martins <andre@cilium.io>
